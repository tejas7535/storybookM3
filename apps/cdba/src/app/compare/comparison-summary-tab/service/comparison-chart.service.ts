import { Injectable } from '@angular/core';

import { TranslocoService } from '@jsverse/transloco';

@Injectable({ providedIn: 'root' })
export class ComparisonChartService {
  constructor(private readonly transloco: TranslocoService) {}

  readonly PURCHASE = this.transloco.translate(
    'compare.summary.common.purchase'
  );
  readonly REMAINDER = this.transloco.translate(
    'compare.summary.common.remainder'
  );
  readonly MOH = this.transloco.translate('compare.summary.common.moh');
  readonly BURDEN = this.transloco.translate('compare.summary.common.burden');
  readonly LABOUR = this.transloco.translate('compare.summary.common.labour');
  readonly RAW_MATERIAL = this.transloco.translate(
    'compare.summary.common.rawMaterial'
  );
  readonly TOTAL = this.transloco.translate('compare.summary.common.total');
}
