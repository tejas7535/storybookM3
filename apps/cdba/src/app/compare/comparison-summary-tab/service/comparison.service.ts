import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';

import { Observable } from 'rxjs';

import { API, ComparisonPath } from '@cdba/shared/constants/api';
import { BomIdentifier } from '@cdba/shared/models';

import { Comparison } from '../../../shared/models/comparison.model';

@Injectable({ providedIn: 'root' })
export class ComparisonService {
  constructor(private readonly httpClient: HttpClient) {}

  getComparison(identifiers: BomIdentifier[]): Observable<Comparison> {
    const path = `${API.v1}/${ComparisonPath}`;

    return this.httpClient.post<Comparison>(path, identifiers);
  }
}
