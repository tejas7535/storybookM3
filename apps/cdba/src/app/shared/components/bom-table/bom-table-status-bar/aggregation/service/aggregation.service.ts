import { Injectable } from '@angular/core';

import { CellRange, GridApi, IRowNode } from 'ag-grid-community';

import { AggregationStatusBar } from '@cdba/shared/models';
import { AggregationStatusBarData } from '@cdba/shared/models/aggregation-status-bar.model';

@Injectable()
export class AggregationService {
  calculateStatusBarValues(
    aggregationModel: AggregationStatusBar,
    cellRanges: CellRange[],
    api: GridApi
  ): AggregationStatusBar {
    if (!cellRanges || cellRanges.length === 0) {
      return this.resetModel(aggregationModel);
    }

    aggregationModel.data = this.extractSelectedCells(cellRanges, api);

    return this.calculateModel(aggregationModel);
  }

  private calculateModel(
    aggregationModel: AggregationStatusBar
  ): AggregationStatusBar {
    let containsNumberCells = false;
    let containsStringCells = false;
    let count = 0;

    if (aggregationModel.data.numberCells.size >= 2) {
      let sum = 0;
      let min = Number.POSITIVE_INFINITY;
      let max = Number.NEGATIVE_INFINITY;
      let average = 0;
      const precision = 4;

      aggregationModel.data.numberCells.forEach((value: number) => {
        if (value < min) {
          min = value;
        }
        if (value > max) {
          max = value;
        }
        count += 1;

        sum += value;
      });

      average = sum / count;

      aggregationModel.average = this.roundUp(average, precision);
      aggregationModel.max = this.roundUp(max, precision);
      aggregationModel.min = this.roundUp(min, precision);
      aggregationModel.sum = this.roundUp(sum, precision);
      containsNumberCells = true;
    } else {
      containsNumberCells = false;
    }
    if (aggregationModel.data.stringCells.size > 0) {
      aggregationModel.data.stringCells.forEach(() => {
        count += 1;
      });
    }
    aggregationModel.count = count;

    containsStringCells = aggregationModel.data.stringCells.size >= 2;

    aggregationModel.isVisible =
      containsNumberCells || containsStringCells ? true : false;

    aggregationModel.showFullModel =
      containsStringCells && !containsNumberCells ? false : true;

    return aggregationModel;
  }

  private extractSelectedCells(
    cellRanges: CellRange[],
    api: GridApi
  ): AggregationStatusBarData {
    const numberCells: Map<string, number> = new Map();
    const stringCells: Map<string, string> = new Map();

    cellRanges.forEach((cellRange) => {
      let cellValue: any;
      let rowNode: IRowNode;

      const rowStartIndex = Math.min(
        cellRange.startRow.rowIndex,
        cellRange.endRow.rowIndex
      );
      const rowEndIndex = Math.max(
        cellRange.startRow.rowIndex,
        cellRange.endRow.rowIndex
      );

      for (let i = rowStartIndex; i <= rowEndIndex; i += 1) {
        cellRange.columns.forEach((column) => {
          rowNode = api.getDisplayedRowAtIndex(i);
          cellValue = api.getCellValue({ rowNode, colKey: column });
          if (typeof cellValue === 'number' && !Number.isNaN(cellValue)) {
            numberCells.set(
              `${column.getId()}_r${rowNode.rowIndex}`,
              cellValue
            );
          } else if (
            typeof cellValue === 'string' &&
            cellValue.trim().length > 0
          ) {
            stringCells.set(
              `${column.getId()}_r${rowNode.rowIndex}`,
              cellValue
            );
          }
        });
      }
    });

    return new AggregationStatusBarData(numberCells, stringCells);
  }

  private resetModel(
    aggregationModel: AggregationStatusBar
  ): AggregationStatusBar {
    aggregationModel.data = new AggregationStatusBarData(new Map(), new Map());
    aggregationModel.average = 0;
    aggregationModel.count = 0;
    aggregationModel.max = 0;
    aggregationModel.min = 0;
    aggregationModel.sum = 0;
    aggregationModel.isVisible = false;
    aggregationModel.showFullModel = false;

    return aggregationModel;
  }

  private roundUp(num: number, precision: number) {
    const prec: number = Math.pow(10, precision);

    return Math.ceil(num * prec) / prec;
  }
}
