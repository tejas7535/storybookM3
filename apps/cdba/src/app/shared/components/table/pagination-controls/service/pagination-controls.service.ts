/* eslint-disable @typescript-eslint/member-ordering */
// cannot have member-ordering with adjacent-overload-signatures rule simultaneously
import { Inject, Injectable } from '@angular/core';

import { LOCAL_STORAGE } from '@ng-web-apis/common';

import {
  MIN_PAGE_SIZE,
  PAGE_SIZE_OPTIONS,
  USER_PAGE_SIZE_KEY,
} from '@cdba/shared/constants/pagination';

@Injectable({ providedIn: 'root' })
export class PaginationControlsService {
  _currentPage = 0;
  _pages: number;
  _range: number;

  constructor(@Inject(LOCAL_STORAGE) private readonly localStorage: Storage) {}

  get currentPage(): number {
    return this._currentPage;
  }
  set currentPage(pageNumber: number) {
    this._currentPage = pageNumber;
  }
  get pages(): number {
    return this._pages;
  }
  set pages(pages: number) {
    this._pages = pages;
  }
  get range(): number {
    return this._range;
  }
  set range(range: number) {
    this._range = range;
  }

  getPageSizeFromLocalStorage(): number {
    const localStoragePageSize = Number(
      this.localStorage.getItem(USER_PAGE_SIZE_KEY)
    );

    if (
      localStoragePageSize !== undefined &&
      PAGE_SIZE_OPTIONS.includes(localStoragePageSize)
    ) {
      return localStoragePageSize;
    } else {
      this.localStorage.setItem(USER_PAGE_SIZE_KEY, String(MIN_PAGE_SIZE));

      return MIN_PAGE_SIZE;
    }
  }

  setPageSizeToLocalStorage(newPageSize: number): void {
    // Ignore values not from available options
    if (PAGE_SIZE_OPTIONS.includes(newPageSize)) {
      this.localStorage.setItem(USER_PAGE_SIZE_KEY, String(newPageSize));
    }
  }
}
