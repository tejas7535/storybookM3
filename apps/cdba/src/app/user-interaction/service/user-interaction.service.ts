import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import {
  MatSnackBar,
  MatSnackBarRef,
  TextOnlySnackBar,
} from '@angular/material/snack-bar';

import { Observable } from 'rxjs';

import { TranslocoService } from '@jsverse/transloco';

import { API, BomExportStatusPath } from '@cdba/shared/constants/api';

import { BomExportStatus } from '../model/feature/bom-export';
import { Interaction } from '../model/interaction.model';
import { InteractionType } from '../model/interaction-type.enum';

@Injectable({ providedIn: 'root' })
export class UserInteractionService {
  snackBarRef: MatSnackBarRef<TextOnlySnackBar>;
  snackBarIsOpen = false;

  public constructor(
    private readonly snackbar: MatSnackBar,
    private readonly translocoService: TranslocoService,
    private readonly http: HttpClient
  ) {}

  /**
   * Opens a SnackBar with the message and action provided by the interactionType
   *
   * @param {InteractionType} interactionType - The type of interaction to handle
   */
  interact(interactionType: InteractionType): void {
    let interaction: Interaction;
    if (!this.snackBarIsOpen) {
      interaction = this.determineInteractionByType(interactionType);

      const message = this.translocoService.translate(`${interaction.message}`);
      let actionMessage = '';
      if (interaction.action) {
        actionMessage = this.translocoService.translate(
          `${interaction.action.message}`
        );
      }

      this.snackBarRef = this.snackbar.open(message, actionMessage);

      this.snackBarIsOpen = true;

      if (interaction.action && interaction.action.url !== '') {
        this.snackBarRef.onAction().subscribe(() => {
          window.open(interaction.action.url, '_blank').focus();
        });
      }

      this.snackBarRef.afterDismissed().subscribe(() => {
        this.snackBarIsOpen = false;
      });
    }
  }

  /**
   * Loads the initial BOM (Bill of Materials) export status.
   *
   * @returns An Observable that emits the current BOM export status.
   */
  loadInitialBomExportStatus(): Observable<BomExportStatus> {
    return this.http.get<BomExportStatus>(`${API.v1}/${BomExportStatusPath}`);
  }

  // Required due to the unserializable action exception from NgRx
  private determineInteractionByType(
    interactionType: InteractionType
  ): Interaction {
    let interaction: Interaction;
    switch (interactionType) {
      case InteractionType.HTTP_GENERAL_ERROR: {
        interaction = Interaction.HTTP_GENERAL_ERROR;
        break;
      }
      case InteractionType.HTTP_GENERAL_VALIDATION_ERROR: {
        interaction = Interaction.HTTP_GENERAL_VALIDATION_ERROR;
        break;
      }
      case InteractionType.REQUEST_BOM_EXPORT_SUCCESS: {
        interaction = Interaction.REQUEST_BOM_EXPORT_SUCCESS;
        break;
      }
      case InteractionType.REQUEST_BOM_EXPORT_FAILURE: {
        interaction = Interaction.REQUEST_BOM_EXPORT_FAILURE;
        break;
      }
      case InteractionType.TRACK_BOM_EXPORT_PROGRESS_COMPLETED: {
        interaction = Interaction.TRACK_BOM_EXPORT_PROGRESS_COMPLETED;
        break;
      }
      case InteractionType.TRACK_BOM_EXPORT_PROGRESS_FAILURE: {
        interaction = Interaction.TRACK_BOM_EXPORT_PROGRESS_FAILURE;
        break;
      }
      case InteractionType.REQUEST_BOM_EXPORT_VALIDATION_ERROR: {
        interaction = Interaction.REQUEST_BOM_EXPORT_VALIDATION_ERROR;
        break;
      }
      default: {
        throw new Error('Unhandled InteractionType');
      }
    }

    return interaction;
  }
}
