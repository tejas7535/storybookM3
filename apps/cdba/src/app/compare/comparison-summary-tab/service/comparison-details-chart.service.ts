import { Injectable } from '@angular/core';

import { TranslocoLocaleService } from '@jsverse/transloco-locale';
import { EChartsOption } from 'echarts';

import { DataPoint } from '@cdba/shared/components/bom-chart/data-point.model';
import { ComparisonType } from '@cdba/shared/constants';
import {
  CHART_HIGHLIGHT_ERROR,
  CHART_PRIMARY,
  CHART_SECONDARY,
  COST_SHARE_CATEGORY_COLORS,
} from '@cdba/shared/constants/colors';
import { ComparisonDetail } from '@cdba/shared/models/comparison.model';
import { CostShareService } from '@cdba/shared/services';

import { ComparisonChartService } from './comparison-chart.service';

@Injectable({ providedIn: 'root' })
export class ComparisonDetailsChartService {
  constructor(
    private readonly comparisonChartService: ComparisonChartService,
    private readonly localeService: TranslocoLocaleService,
    private readonly costShareService: CostShareService
  ) {}

  provideParetoChartConfig(
    details: ComparisonDetail[],
    currency: string
  ): EChartsOption {
    const xAxisData: string[] = [];
    const yAxisLineData: number[] = [];
    const yAxisBarDataPoints: DataPoint[] = [];
    let min = Number.POSITIVE_INFINITY;
    let max = Number.NEGATIVE_INFINITY;
    let cumulativePercentageCost = 0;

    details.forEach((detail) => {
      if (detail.costSharePercentage !== 0) {
        xAxisData.push(detail.title);
        min = Math.min(min, detail.totalCosts);
        max = Math.max(max, detail.totalCosts);

        cumulativePercentageCost += detail.costSharePercentage;
        cumulativePercentageCost = Number.parseFloat(
          cumulativePercentageCost.toPrecision(4)
        );
        yAxisLineData.push(cumulativePercentageCost);
        yAxisBarDataPoints.push({
          name: detail.title,
          value: detail.totalCosts,
          itemStyle: {
            color: COST_SHARE_CATEGORY_COLORS.get(
              this.costShareService.getCostShareCategory(
                detail.costSharePercentage / 100
              )
            ),
          },
        });
      }
    });

    return {
      tooltip: {
        trigger: 'axis',
        axisPointer: {
          type: 'cross',
          crossStyle: {
            color: '#999',
          },
        },
      },
      toolbox: {
        feature: {
          saveAsImage: { show: true },
        },
      },
      xAxis: [
        {
          type: 'category',
          data: xAxisData,
          axisPointer: {
            type: 'shadow',
          },
        },
      ],
      yAxis: [
        {
          type: 'value',
          min,
          max,
          axisLabel: {
            formatter: (value: string) => `${value} ${currency}`,
          },
        },
        {
          type: 'value',
          min: 0,
          max: 100,
          interval: 10,
          axisLabel: {
            formatter: (value: string) => `${value} %`,
          },
        },
      ],
      series: [
        {
          type: 'bar',
          tooltip: {
            valueFormatter: (value: number) =>
              `${this.localeService.localizeNumber(value, 'decimal')} ${currency}`,
          },
          data: yAxisBarDataPoints,
        },
        {
          type: 'line',
          color: CHART_HIGHLIGHT_ERROR,
          yAxisIndex: 1,
          tooltip: {
            valueFormatter: (value: string) => `${value} %`,
          },
          data: yAxisLineData,
        },
      ],
    } as EChartsOption;
  }

  provideDetailedBarChartConfig(
    detail: ComparisonDetail,
    firstMaterialDesignation: string,
    secondMaterialDesignation: string
  ): EChartsOption {
    const firstMaterialDesignationCosts: number[] = [];
    const secondMaterialDesignationCosts: number[] = [];

    detail.costDifferences.forEach((costDifference) => {
      switch (costDifference.type) {
        case ComparisonType.PURCHASE: {
          firstMaterialDesignationCosts[0] = costDifference.valueBom1;
          secondMaterialDesignationCosts[0] = costDifference.valueBom2;
          break;
        }
        case ComparisonType.RAW_MATERIAL: {
          firstMaterialDesignationCosts[1] = costDifference.valueBom1;
          secondMaterialDesignationCosts[1] = costDifference.valueBom2;
          break;
        }
        case ComparisonType.LABOUR: {
          firstMaterialDesignationCosts[2] = costDifference.valueBom1;
          secondMaterialDesignationCosts[2] = costDifference.valueBom2;
          break;
        }
        case ComparisonType.BURDEN: {
          firstMaterialDesignationCosts[3] = costDifference.valueBom1;
          secondMaterialDesignationCosts[3] = costDifference.valueBom2;
          break;
        }
        case ComparisonType.MOH: {
          firstMaterialDesignationCosts[4] = costDifference.valueBom1;
          secondMaterialDesignationCosts[4] = costDifference.valueBom2;
          break;
        }
        case ComparisonType.REMAINDER: {
          firstMaterialDesignationCosts[5] = costDifference.valueBom1;
          secondMaterialDesignationCosts[5] = costDifference.valueBom2;
          break;
        }
        case ComparisonType.TOTAL: {
          firstMaterialDesignationCosts[6] = costDifference.valueBom1;
          secondMaterialDesignationCosts[6] = costDifference.valueBom2;
          break;
        }
        default: {
          throw new Error(`Invalid Comparison Type: ${costDifference.type}`);
        }
      }
    });

    return {
      tooltip: {
        trigger: 'axis',
        axisPointer: {
          type: 'shadow',
        },
      },
      xAxis: {
        type: 'category',
        data: [
          this.comparisonChartService.PURCHASE,
          this.comparisonChartService.RAW_MATERIAL,
          this.comparisonChartService.LABOUR,
          this.comparisonChartService.BURDEN,
          this.comparisonChartService.MOH,
          this.comparisonChartService.REMAINDER,
          this.comparisonChartService.TOTAL,
        ],
        axisLabel: {
          interval: 0,
          width: 200,
        },
      },
      yAxis: {
        type: 'value',
      },
      series: [
        {
          type: 'bar',
          name: firstMaterialDesignation,
          data: firstMaterialDesignationCosts,
          itemStyle: {
            color: CHART_PRIMARY,
          },
        },
        {
          type: 'bar',
          name: secondMaterialDesignation,
          data: secondMaterialDesignationCosts,
          itemStyle: {
            color: CHART_SECONDARY,
          },
        },
      ],
    } as EChartsOption;
  }
}
