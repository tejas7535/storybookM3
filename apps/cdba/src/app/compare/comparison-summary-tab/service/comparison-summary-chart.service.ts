import { Injectable } from '@angular/core';

import { TranslocoService } from '@jsverse/transloco';
import { EChartsOption } from 'echarts/types/dist/echarts';

import {
  CHART_PRIMARY,
  CHART_SECONDARY,
  RADAR_CHART_REFERENCE_LINE,
} from '@cdba/shared/constants/colors';
import { ComparisonType } from '@cdba/shared/constants/comparison-type';
import {
  Comparison,
  ComparisonSummary,
  CostDifference,
} from '@cdba/shared/models/comparison.model';

import { ComparisonChartService } from './comparison-chart.service';

@Injectable({ providedIn: 'root' })
export class ComparisonSummaryChartService {
  constructor(
    private readonly transloco: TranslocoService,
    private readonly comparisonChartService: ComparisonChartService
  ) {}

  provideSummaryTableColumns(): string[] {
    return [
      'costType',
      'valueBom1',
      'valueBom2',
      'currency',
      'costDeltaValue',
      'costDeltaPercentage',
    ];
  }

  provideSummarizedRadarChartConfig(
    firstMaterialDesignation: string,
    secondMaterialDesignation: string,
    comparison: Comparison
  ): EChartsOption {
    const referenceLineTitle = this.transloco.translate(
      'compare.summary.common.referenceLine'
    );

    const purchaseCost = this.findCostDifferenceByComparisonType(
      comparison.summary,
      ComparisonType.PURCHASE
    );
    const remainderCost = this.findCostDifferenceByComparisonType(
      comparison.summary,
      ComparisonType.REMAINDER
    );
    const mohCost = this.findCostDifferenceByComparisonType(
      comparison.summary,
      ComparisonType.MOH
    );
    const burdenCost = this.findCostDifferenceByComparisonType(
      comparison.summary,
      ComparisonType.BURDEN
    );
    const labourCost = this.findCostDifferenceByComparisonType(
      comparison.summary,
      ComparisonType.LABOUR
    );
    const rawMaterialCost = this.findCostDifferenceByComparisonType(
      comparison.summary,
      ComparisonType.RAW_MATERIAL
    );

    // Calculate min max values
    let min = Number.POSITIVE_INFINITY;
    let max = Number.NEGATIVE_INFINITY;
    comparison.summary.costDifferences.forEach((costDifference) => {
      if (costDifference.type !== 'TOTAL') {
        min = Math.min(min, costDifference.valueBom1, costDifference.valueBom2);
        max = Math.max(max, costDifference.valueBom1, costDifference.valueBom2);
      }
    });
    min = Math.floor(min);
    max = Math.ceil(max);

    return {
      legend: {
        bottom: '0%',
        left: '5%',
        data: [
          referenceLineTitle,
          firstMaterialDesignation,
          secondMaterialDesignation,
        ],
      },
      tooltip: {
        trigger: 'item',
        valueFormatter: (value: string) => `${value} ${comparison.currency}`,
      },
      toolbox: {
        feature: {
          saveAsImage: { show: true },
        },
      },
      radar: {
        indicator: [
          {
            name: this.comparisonChartService.PURCHASE,
            min,
            max,
          },
          {
            name: this.comparisonChartService.REMAINDER,
            min,
            max,
          },
          {
            name: this.comparisonChartService.MOH,
            min,
            max,
          },
          {
            name: this.comparisonChartService.BURDEN,
            min,
            max,
          },
          {
            name: this.comparisonChartService.LABOUR,
            min,
            max,
          },
          {
            name: this.comparisonChartService.RAW_MATERIAL,
            min,
            max,
          },
        ],
      },
      series: [
        {
          type: 'radar',
          data: [
            {
              name: referenceLineTitle,
              value: [0, 0, 0, 0, 0, 0],
              itemStyle: {
                color: RADAR_CHART_REFERENCE_LINE,
              },
            },
            {
              name: firstMaterialDesignation,
              value: [
                purchaseCost.valueBom1,
                remainderCost.valueBom1,
                mohCost.valueBom1,
                burdenCost.valueBom1,
                labourCost.valueBom1,
                rawMaterialCost.valueBom1,
              ],
              itemStyle: {
                color: CHART_PRIMARY,
              },
            },
            {
              name: secondMaterialDesignation,
              value: [
                purchaseCost.valueBom2,
                remainderCost.valueBom2,
                mohCost.valueBom2,
                burdenCost.valueBom2,
                labourCost.valueBom2,
                rawMaterialCost.valueBom2,
              ],
              itemStyle: {
                color: CHART_SECONDARY,
              },
            },
          ],
        },
      ],
    } as EChartsOption;
  }

  findCostDifferenceByComparisonType(
    summary: ComparisonSummary,
    type: ComparisonType
  ): CostDifference {
    return summary.costDifferences.find(
      (costDifference) => costDifference.type === type
    );
  }
}
