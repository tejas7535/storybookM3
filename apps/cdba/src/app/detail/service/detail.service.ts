import { HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';

import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';

import { DataService } from '../../core/http/data.service';
import {
  BomIdentifier,
  BomItem,
  BomResult,
  ReferenceTypeIdModel,
  ReferenceTypeResultModel,
} from '../../core/store/reducers/detail/models';
import { CalculationsResultModel } from '../../core/store/reducers/detail/models/calculations-result-model';
import { Calculation } from '../../core/store/reducers/shared/models/calculation.model';

@Injectable({
  providedIn: 'root',
})
export class DetailService {
  private readonly DETAIL = 'detail';

  private readonly CALCULATIONS = 'calculations';

  private readonly PARAM_MATERIAL_NUMBER = 'material-number';
  private readonly PARAM_PLANT = 'plant';

  private readonly BOM = 'bom';
  private readonly PARAM_BOM_COSTING_DATE = 'bom-costing-date';
  private readonly PARAM_BOM_COSTING_NUMBER = 'bom-costing-number';
  private readonly PARAM_BOM_COSTING_TYPE = 'bom-costing-type';
  private readonly PARAM_BOM_COSTING_VERSION = 'bom-costing-version';
  private readonly PARAM_BOM_ENTERED_MANUALLY = 'bom-entered-manually';
  private readonly PARAM_BOM_REFERENCE_OBJECT = 'bom-reference-object';
  private readonly PARAM_BOM_VALUATION_VARIANT = 'bom-valuation-variant';

  private static defineBomTreeForAgGrid(
    items: BomItem[],
    idx: number
  ): BomItem[] {
    if (idx === items.length) {
      return items;
    }

    const currentItem = items[idx];
    const nextIdx = idx + 1;

    // special case level 1
    if (currentItem.level === 1) {
      currentItem.predecessorsInTree = [currentItem.materialDesignation];

      return DetailService.defineBomTreeForAgGrid(items, nextIdx);
    }

    const previousItem = items[idx - 1];

    if (previousItem.level < currentItem.level) {
      // previous item is parent
      currentItem.predecessorsInTree = [
        ...previousItem.predecessorsInTree,
        currentItem.materialDesignation,
      ];
    } else {
      // previous item has same parent or different parent
      const itemsToRemove = previousItem.level - currentItem.level + 1;
      const predecessors = [...previousItem.predecessorsInTree];

      // remove last items to have the actual parent as last item
      predecessors.splice(predecessors.length - itemsToRemove, itemsToRemove);

      // since duplicates are possible, add spaces to the key until it is unique
      let foundUniqueKey = false;
      let currentMaterialDesignation = currentItem.materialDesignation;
      while (!foundUniqueKey) {
        const temp = [...predecessors, currentMaterialDesignation];

        if (
          items
            .map((item) => item.predecessorsInTree)
            .find(
              (materialDesignations: string[]) =>
                materialDesignations.length === temp.length &&
                materialDesignations.every(
                  (value: string, index: number) => value === temp[index]
                )
            ) !== undefined
        ) {
          currentMaterialDesignation += ' ';
        } else {
          foundUniqueKey = true;
        }
      }

      currentItem.predecessorsInTree = [
        ...predecessors,
        currentMaterialDesignation,
      ];
    }

    return DetailService.defineBomTreeForAgGrid(items, nextIdx);
  }

  public constructor(private readonly dataService: DataService) {}

  public detail(
    item: ReferenceTypeIdModel
  ): Observable<ReferenceTypeResultModel> {
    const params: HttpParams = new HttpParams()
      .set(this.PARAM_MATERIAL_NUMBER, item.materialNumber)
      .set(this.PARAM_PLANT, item.plant);

    return this.dataService.getAll<ReferenceTypeResultModel>(
      this.DETAIL,
      params
    );
  }

  public calculations(materialNumber: string): Observable<Calculation[]> {
    const params = new HttpParams().set(
      this.PARAM_MATERIAL_NUMBER,
      materialNumber
    );

    return this.dataService
      .getAll<CalculationsResultModel>(this.CALCULATIONS, params)
      .pipe(map((response: CalculationsResultModel) => response.items));
  }

  public getBom(bomIdentifier: BomIdentifier): Observable<BomItem[]> {
    const params = new HttpParams()
      .set(this.PARAM_BOM_COSTING_DATE, bomIdentifier.bomCostingDate)
      .set(this.PARAM_BOM_COSTING_NUMBER, bomIdentifier.bomCostingNumber)
      .set(this.PARAM_BOM_COSTING_TYPE, bomIdentifier.bomCostingType)
      .set(this.PARAM_BOM_COSTING_VERSION, bomIdentifier.bomCostingVersion)
      .set(this.PARAM_BOM_ENTERED_MANUALLY, bomIdentifier.bomEnteredManually)
      .set(this.PARAM_BOM_REFERENCE_OBJECT, bomIdentifier.bomReferenceObject)
      .set(this.PARAM_BOM_VALUATION_VARIANT, bomIdentifier.bomValuationVariant);

    return this.dataService.getAll<BomResult>(this.BOM, params).pipe(
      map((response: BomResult) =>
        response.items.map((item) => ({
          ...item,
          predecessorsInTree: [],
        }))
      ),
      map((items: BomItem[]) => DetailService.defineBomTreeForAgGrid(items, 0))
    );
  }
}
