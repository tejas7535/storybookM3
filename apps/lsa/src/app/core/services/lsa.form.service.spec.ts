import { FormGroup, Validators } from '@angular/forms';

import {
  ApplicationFormValue,
  LubricantFormValue,
  LubricationPointsFormValue,
  RecommendationFormValue,
} from '@lsa/shared/models';
import { createServiceFactory, SpectatorService } from '@ngneat/spectator/jest';

import { LsaFormService } from './lsa-form.service';

describe('LsaFormService', () => {
  let spectator: SpectatorService<LsaFormService>;
  let service: LsaFormService;

  const createService = createServiceFactory(LsaFormService);

  beforeEach(() => {
    spectator = createService();
    service = spectator.service;
  });

  it('should create', () => {
    expect(service).toBeTruthy();
  });

  describe('isValid', () => {
    it('should return form valid', () => {
      expect(service.isValid).toBe(false);

      service.recommendationForm = { valid: true } as FormGroup;

      expect(service.isValid).toBe(true);
    });
  });

  describe('getLubricationPointsForm', () => {
    it('should return the lubricationPointsForm', () => {
      const mockForm = {} as FormGroup;
      service.lubricationPointsForm = mockForm;

      const result = service.getLubricationPointsForm();

      expect(result).toEqual(mockForm);
    });
  });

  describe('updateLubricationPointsForm', () => {
    it('should patch the lubricationPointsForm', () => {
      const mockValue = {} as LubricationPointsFormValue;
      service.lubricationPointsForm.patchValue = jest.fn();

      service.updateLubricationPointsForm(mockValue);

      expect(service.lubricationPointsForm.patchValue).toHaveBeenCalledWith(
        mockValue,
        { emitEvent: undefined, onlySelf: true }
      );
    });
  });

  describe('getLubricantForm', () => {
    it('should return the lubricantForm', () => {
      const mockForm = {} as FormGroup;
      service.lubricantForm = mockForm;

      const result = service.getLubricantForm();

      expect(result).toEqual(mockForm);
    });
  });

  describe('updateLubricantForm', () => {
    it('should patch the lubricantForm', () => {
      const mockValue = {} as LubricantFormValue;
      service.lubricantForm.patchValue = jest.fn();

      service.updateLubricantForm(mockValue);

      expect(service.lubricantForm.patchValue).toHaveBeenCalledWith(mockValue, {
        emitEvent: undefined,
        onlySelf: true,
      });
    });
  });

  describe('getApplicationForm', () => {
    it('should return the applicationForm', () => {
      const mockForm = {} as FormGroup;
      service.applicationForm = mockForm;

      const result = service.getApplicationForm();

      expect(result).toEqual(mockForm);
    });
  });

  describe('updateApplicationForm', () => {
    it('should patch the applicationForm', () => {
      const mockValue = {} as ApplicationFormValue;
      service.applicationForm.patchValue = jest.fn();

      service.updateApplicationForm(mockValue);

      expect(service.applicationForm.patchValue).toHaveBeenCalledWith(
        mockValue,
        { emitEvent: undefined, onlySelf: true }
      );
    });
  });

  describe('getRecommendationForm', () => {
    it('should return the recommendationForm', () => {
      const mockForm = {} as FormGroup;
      service.recommendationForm = mockForm;

      const result = service.getRecommendationForm();

      expect(result).toEqual(mockForm);
    });
  });

  describe('updateRecommendationForm', () => {
    it('should patch the recommendationForm', () => {
      const mockValue = {} as RecommendationFormValue;
      service.recommendationForm.patchValue = jest.fn();

      service.updateRecommendationForm(mockValue);

      expect(service.recommendationForm.patchValue).toHaveBeenCalledWith(
        mockValue
      );
    });
  });

  describe('init', () => {
    it('should init the form with defaults', () => {
      service['initForm']();

      expect(service.lubricationPointsForm.getRawValue()).toMatchSnapshot();
      expect(service.lubricantForm.getRawValue()).toMatchSnapshot();
      expect(service.applicationForm.getRawValue()).toMatchSnapshot();
      expect(service.recommendationForm.getRawValue()).toMatchSnapshot();
    });
  });

  describe('createFormControl', () => {
    it('should return a form control', () => {
      const result = service['createFormControl']<string>('value');

      expect(result.value).toEqual('value');
      expect(result.hasValidator(Validators.required)).toBe(false);
    });

    it('should return a required form control', () => {
      const result = service['createFormControl']<string>('value', true);

      expect(result.value).toEqual('value');
      expect(result.hasValidator(Validators.required)).toBe(true);
    });
  });
});
