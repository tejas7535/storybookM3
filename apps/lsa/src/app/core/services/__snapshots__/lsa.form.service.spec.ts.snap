// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LsaFormService init should init the form with defaults 1`] = `
Object {
  "lubricationInterval": "year",
  "lubricationPoints": "2-4",
  "lubricationQty": 60,
  "optime": 1,
  "pipeLength": Object {
    "max": 3,
    "min": 1,
    "title": "1 - 3m",
  },
}
`;

exports[`LsaFormService init should init the form with defaults 2`] = `
Object {
  "grease": Object {
    "id": "ARCANOL_MULTI2",
    "title": "Arcanol MULTI2",
  },
  "lubricantType": "Arcanol",
}
`;

exports[`LsaFormService init should init the form with defaults 3`] = `
Object {
  "battery": 2,
  "temperature": null,
}
`;

exports[`LsaFormService init should init the form with defaults 4`] = `
Object {
  "application": Object {
    "battery": 2,
    "temperature": null,
  },
  "lubricant": Object {
    "grease": Object {
      "id": "ARCANOL_MULTI2",
      "title": "Arcanol MULTI2",
    },
    "lubricantType": "Arcanol",
  },
  "lubricationPoints": Object {
    "lubricationInterval": "year",
    "lubricationPoints": "2-4",
    "lubricationQty": 60,
    "optime": 1,
    "pipeLength": Object {
      "max": 3,
      "min": 1,
      "title": "1 - 3m",
    },
  },
}
`;
